core/state_transition.go

```diff
@@ -163,12 +165,64 @@ func IntrinsicGas(data []byte, accessList types.AccessList, isContractCreation b
 		gas += z * params.TxDataZeroGas
 	}
 	if accessList != nil {
+		accessListGas, err := AccessListGas(rules, accessList)
+		if err != nil {
+			return 0, err
+		}
+		totalGas, overflow := commonMath.SafeAdd(gas, accessListGas)
+		if overflow {
+			return 0, fmt.Errorf("overflow adding access list gas")
+		}
+		gas = totalGas
+	}
+
+	return gas, nil
+}
+
+func AccessListGas(rules params.Rules, accessList types.AccessList) (uint64, error) {
+	var gas uint64
+	if !rules.PredicatesExist() {
 		gas += uint64(len(accessList)) * params.TxAccessListAddressGas
 		gas += uint64(accessList.StorageKeys()) * params.TxAccessListStorageKeyGas
+		return gas, nil
 	}
+
+	for i, accessTuple := range accessList {
+		address := accessTuple.Address
+		if !rules.PredicateExists(address) {
+			totalGas, overflow := commonMath.SafeAdd(gas, params.TxAccessListAddressGas+uint64(len(accessTuple.StorageKeys))*params.TxAccessListStorageKeyGas)
+			if overflow {
+				return 0, fmt.Errorf("overflow adding storage slot gas at index %d", i)
+			}
+			gas = totalGas
+		} else {
+			predicateGas, err := ApplyPredicateGas(rules, accessTuple)
+			if err != nil {
+				return 0, err
+			}
+			totalGas, overflow := commonMath.SafeAdd(gas, predicateGas)
+			if overflow {
+				return 0, fmt.Errorf("overflow adding predicate gas at index %d", i)
+			}
+			gas = totalGas
+		}
+	}
+
 	return gas, nil
 }
 
+func ApplyPredicateGas(rules params.Rules, accessTuple types.AccessTuple) (uint64, error) {
+	predicate, ok := rules.PredicatePrecompiles[accessTuple.Address]
+	if ok {
+		return predicate.PredicateGas(utils.HashSliceToBytes(accessTuple.StorageKeys))
+	}
+	proposerPredicates, ok := rules.ProposerPredicates[accessTuple.Address]
+	if !ok {
```

Comment:
nit: could we re-write as above for symmetry?

Reply:
Removing the `s`